### PR TITLE
Feat: Optimistic fast finality

### PR DESCRIPTION
**Pull request by phuctd95**

### CODE REVIEW

File: .github/workflows/test-pr.yml
Comment (minh-bq) [inline .github/workflows/test-pr.yml:1]: We need to revert this change when the PR is ready to merge to master

**Review by minh-bq — CHANGES_REQUESTED** [review]

File: consensus/consortium/v2/consortium.go [hunk 1]
```diff
@@ -570,15 +640,37 @@ func (c *Consortium) Prepare(chain consensus.ChainHeaderReader, header *types.He
...
+			blsPublicKeys       []*blst.PublicKey = make([]*blst.PublicKey, len(newValidators))
+			checkpointValidator []finality.ValidatorWithBlsPub
+		)
+
+		// Before Shillin, when encoding the checkpointExtra into header's extra data,
+		// we ignore the blsPublicKeys. So just leave the checkpointExtra's blsPublicKeys
+		// filled with zero value before Shillin.
+		if isShillin {
+			// TODO: Get finality voter public key from contract
+			// blsPublicKeys =
```
Comment (minh-bq) [inline consensus/consortium/v2/consortium.go:654]: TODO: We need to filter out the validator without BLS public key or with a malformed one

File: core/blockchain.go [hunk 2]
```diff
@@ -1279,6 +1323,38 @@ func (bc *BlockChain) WriteBlockWithState(block *types.Block, receipts []*types.
...
+func (bc *BlockChain) reorgNeeded(localBlock *types.Block, localTd *big.Int, externBlock *types.Block, externTd *big.Int) bool {
+	if consensusEngine, ok := bc.engine.(consensus.FastFinalityPoSA); ok {
+		localJustifiedBlockNumber, _ := consensusEngine.GetJustifiedBlock(bc, localBlock.NumberU64(), localBlock.Hash())
+		externJustifiedBlockNumber, _ := consensusEngine.GetJustifiedBlock(bc, externBlock.NumberU64(), externBlock.Hash())
+
+		if externJustifiedBlockNumber > localJustifiedBlockNumber {
+			return true
+		}
+	}
+
```
Comment (phuctd95) [inline core/blockchain.go:1336]: do we need to add: if externJustifiedBlockNumber < localJustifiedBlockNumber, return false?

Comment (minh-bq) [reply core/blockchain.go:1336]: Yes, I will fix this

File: consensus/consortium/v2/consortium.go [hunk 3]
```diff
@@ -200,6 +206,67 @@ func (c *Consortium) GetRecents(chain consensus.ChainHeaderReader, number uint64
...
+) error {
+	snap, err := c.snapshot(chain, parentNumber, parentHash, parents)
+	if err != nil {
+		return err
+	}
+
+	votedValidatorPositions := finalityVotedValidators.Indices()
+	if len(votedValidatorPositions) <= int(math.Floor(finalityRatio*float64(len(snap.ValidatorsWithBlsPub))))+1 {
+		return finality.ErrNotEnoughFinalityVote
+	}
```
Comment (phuctd95) [inline consensus/consortium/v2/consortium.go:247]: I think it should < instead of <=

**Review by minh-bq — APPROVED** [review]
